File: leetcode_ts/longest_common_prefix.ts
function longestCommonPrefix(strs: string[]): string {
    if (strs.length == 0) {
        return "";
    }

    let commonPrefix = '';

    let cursor = 0;

    while (true) {
        let curChar = strs[0]![cursor];

        if (curChar == undefined) {
            return commonPrefix;
        }

        for (let i = 0; i < strs.length; i++) {
            if ((strs[i]?.length ?? 0) <= cursor || strs[i]![cursor] != curChar) {
                return commonPrefix;                                                                            
            }
        }
        commonPrefix += curChar;

        cursor++;
    }
};

let strs = ["flower", "flow", "flight"];
console.log(longestCommonPrefix(strs));

File: leetcode/longest_common_prefix.go
package main

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	curCursor := -1
	commonPrefix := ""
	for {
		curCursor++

		if curCursor >= len(strs[0]) {
			return commonPrefix
		}

		isAllSame := true
		curChar := strs[0][curCursor]
		for i, _ := range strs {
			if curCursor >= len(strs[i]) {
				isAllSame = false
				break
			}

			if strs[i][curCursor] != curChar {
				isAllSame = false
				break
			}
		}
		if isAllSame {
			commonPrefix += string(curChar)
		} else {
			return commonPrefix
		}
	}

}
